Implementation of TraceKVStore

alexanderbez opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺ 
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes. 
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  --> 
#### Changes
* Implemented `TraceKVStore`
* Cleaned up some godocs (mostly server, stores and baseapp)
* Cleaned up structure in server constructors
* Added a CLI flag for KVStore tracing output (such a flag dictates tracing is desired opposed to needing two separate flags)
* Updated <root/cache>MultStore to support store tracing

#### TODO
* [x] How to write the actual keys/values? Right now they're stringified (@cwgoes)
* [x] Use TraceKVStore "under" a `CacheKVStore`, otherwise, all operations are traced regardless if they're in the cache or not (this will severely bloat up the trace log).
* [x] Rebase onto latest `develop`

<hr>

* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG.md
* [ ] Updated Gaia/Examples
* [x] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))
* [x] Added appropriate labels to PR (ex. wip, ready-for-review, docs)

closes: #1523

cwgoes commented on store/tracekvstore.go:
```diff
@@ -0,0 +1,128 @@
+
+	// traceOperation implements a traced KVStore operation
+	traceOperation struct {
+		Timestamp int       `json:"timestamp,omitempty"`
```
I think we want to log block height, not timestamp.

alexanderbez replied:
@cwgoes suggestions about how to go about doing this? Do we need to pass/use the context?

cwgoes commented on store/tracekvstore.go:
```diff
@@ -0,0 +1,128 @@
+
+// iterator facilitates iteration over a KVStore. It delegates the necessary
+// calls to it's parent KVStore.
+func (tkv *TraceKVStore) iterator(start, end []byte, ascending bool) sdk.Iterator {
```
We want to log operations conducted through iterators too; probably need a `TraceKVStoreIterator` wrapper, see `gasIterator` in `store/gaskvstore.go` which is comparable.

cwgoes commented on store/tracekvstore.go:
```diff
@@ -0,0 +1,128 @@
+		Value:     string(value),
+	})
+
+	if err == nil {
```
Does this fail sometimes? Might be better to panic.

alexanderbez replied:
Mhmmm, I thought about that. It should never panic really. Do we really wanna panic on a failure to trace some operation? It would make debugging tricky maybe, but idk if it warrants a panic.

cwgoes replied:
I think we should; better than to fail silently (maybe the machine ran out of RAM? Hard to imagine a case when this would panic). This is an opt-in setting anyways; most users won't use a `TraceKVStore`.

alexanderbez replied:
Fair point 👍

cwgoes commented:
Re: how to write keys/values: probably base64 (should be default in JSON if you encode bytes), I'd expect just converting to string would produce a bunch of unusual characters. Means we have to convert to base64 before searching but that's probably unavoidable.

alexanderbez commented:
Updated @cwgoes, please let me know your thoughts.

* Add table-driven unit tests
* Wrapped TraceKVStore behind a cache wrapper
* Added block height to trace context via `BeginBlock`. I tried other means, but ran into troubles, although I'm sure it's possible. 

If you also have an inclination into how to introduce TX hashes and message indexes, please let me know.

cwgoes commented on store/tracekvstore.go:
```diff
@@ -0,0 +1,200 @@
+
+// Next implements the Iterator interface. It will write a traced operation
+// on a previously set key/value pair.
+func (ti *traceIterator) Next() {
```
Hmm, I don't think this implementation is quite what we want - if I create an iterator, read one key, and never call `.Next()`, we won't log any operations - perhaps instead log key and value the first time `.Key()` is called, and reset a boolean on `.Next()`?

cwgoes commented on baseapp/baseapp.go:
```diff
@@ -392,23 +406,29 @@ func handleQueryP2P(app *BaseApp, path []string, req abci.RequestQuery) (res abc
-	// Otherwise app.deliverState will be nil, since it
-	// is reset on Commit.
+	if app.cms.TracingEnabled() {
+		app.cms.WithTracingContext(&sdk.TraceContext{
```
To add transaction hashes, add `TxHash` to `sdk.TraceContext` and set it in `runTx`.

alexanderbez commented:
@cwgoes Updated to include TX hashes. I changed to the context stuff to be a map so it can be more flexible. I also updated the tracing to trace individual `key`/`value` calls.

re: TX hashes. Probably warrants creating a separate issue for this, but can we update the `sdk.Tx` interface to also include?:
```golang
type Tx struct {
    // ...
    Hash sdk.Hash // sdk.Hash should implement marshaling, Stringer, etc...
}
```

cwgoes commented:
> re: TX hashes. Probably warrants creating a separate issue for this, but can we update the sdk.Tx interface to also include?:

Maybe; if we aren't logging it I'm not sure we need/want to compute the hash, but worth discussing (as a separate issue from this PR).

cwgoes commented on types/store.go:
```diff
@@ -298,3 +316,15 @@ func (getter PrefixStoreGetter) KVStore(ctx Context) KVStore {
+
+// TraceContext contains TraceKVStore context data. It will be written with
+// every trace operation.
+type TraceContext struct {
```
Can this just be `Metadata` (type aliased) - any reason for a wrapper struct?

cwgoes commented on store/rootmultistore.go:
```diff
@@ -130,6 +133,40 @@ func (rs *rootMultiStore) LoadVersion(ver int64) error {
+	return rs
+}
+
+// WithTracingContext updates the tracing context for the MultiStore by merging
```
Update comment to indicate that the metadata will be merged by-key.

alexanderbez commented:
@cwgoes  Updated to use type alias and added examples (including a jq example) to `docs/clients/node.md`

jolesbi commented:
all looks well on the markdown front. not sure if we need to reference tendermint for running a full node but also know that that was not @alexanderbez 's contribution.

👍👍👍
